File: go/internal/sol/p0014/p0014.go
package p0014

import (
	"strconv"
)

func getCollatzLength(n int) int {
	var cnt = 1
	for n > 1 {
		cnt++
		if n&1 == 0 {
			n /= 2
		} else {
			n = 3*n + 1
		}
	}

	return cnt
}

func compute(limit int) string {
	var cnt int
	var maxCnt int
	var answer int
	for i := limit / 2; i < limit; i++ {
		if cnt = getCollatzLength(i); cnt > maxCnt {
			maxCnt = cnt
			answer = i
		}
	}

	return strconv.FormatInt(int64(answer), 10)
}

func Solve() string {
	return compute(1_000_000)
}

File: haskell/src/Sol/P0014.hs
module Sol.P0014 (compute, solve) where

import Control.Arrow ((&&&))
import Data.Function (on)
import Data.List (maximumBy)

-- It is a straightforward method that doesn't use memoization.
getCollatzLength :: Int -> Int
getCollatzLength n = aux n 1
  where
    aux :: Int -> Int -> Int
    aux 1 cnt = cnt
    aux x cnt
        | even x = aux (x `div` 2) (succ cnt)
        | otherwise = aux (3 * x + 1) (succ cnt)

compute :: Int -> String
compute limit =
    show
        . snd
        . maximumBy (compare `on` fst)
        $ (getCollatzLength &&& id) <$> [(div limit 2) .. (limit - 1)]

solve :: String
solve = compute 1_000_000
